server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java

```diff
@@ -0,0 +1,189 @@
+            Setting.Property.NodeScope);
+    public static final Setting<TimeValue> SLOW_PATH_LOGGING_THRESHOLD_SETTING =
+        Setting.timeSetting("monitor.fs.health.slow_path_logging_threshold", TimeValue.timeValueSeconds(5), TimeValue.timeValueMillis(1),
+            Setting.Property.NodeScope, Setting.Property.Dynamic);
+
+
+    public FsHealthService(Settings settings, ClusterSettings clusterSettings, ThreadPool threadPool, NodeEnvironment nodeEnv) {
+        this.threadPool = threadPool;
+        this.enabled = ENABLED_SETTING.get(settings);
+        this.refreshInterval = REFRESH_INTERVAL_SETTING.get(settings);
+        this.slowPathLoggingThreshold = SLOW_PATH_LOGGING_THRESHOLD_SETTING.get(settings);
+        this.currentTimeMillisSupplier = threadPool::relativeTimeInMillis;
+        this.nodeEnv = nodeEnv;
+        this.pathHealthStats = new HashMap<>(1);
+        clusterSettings.addSettingsUpdateConsumer(SLOW_PATH_LOGGING_THRESHOLD_SETTING, this::setSlowPathLoggingThreshold);
+        clusterSettings.addSettingsUpdateConsumer(ENABLED_SETTING, this::setEnabled);
+    }
+
+    @Override
+    protected void doStart() {
+        scheduledFuture = threadPool.scheduleWithFixedDelay(new FsHealthMonitor(), refreshInterval,
+                ThreadPool.Names.GENERIC);
+    }
+
+
+    @Override
+    protected void doStop() {
+        scheduledFuture.cancel();
+        slowPathHealthCheckTimeoutHandler.cancel();
+    }
+
+    @Override
+    protected void doClose() throws IOException {
+
+    }
+
+    public void setEnabled(boolean enabled) { this.enabled = enabled; }
+
+    public void setSlowPathLoggingThreshold(TimeValue slowPathLoggingThreshold) {
+        this.slowPathLoggingThreshold = slowPathLoggingThreshold;
+    }
+
+    @Override
+    public Status getHealth() {
+        if (enabled == false) {
+            return Status.HEALTHY;
+        }
+        return pathHealthStats.entrySet().stream().anyMatch(map -> map.getValue() == Status.UNHEALTHY) ? Status.UNHEALTHY
+            : Status.HEALTHY;
+    }
+
+     class FsHealthMonitor implements Runnable {
+
+        static final String TEMP_FILE_NAME = ".es_temp_file";
+        private byte[] byteToWrite;
+
+        FsHealthMonitor(){
+            this.byteToWrite = new byte[20];
+        }
+
+        @Override
+        public void run() {
+            try {
+                if (enabled) {
+                    monitorFSHealth();
+                    logger.debug("Monitor for FS health ran successfully {}", pathHealthStats);
+                }
+            } catch (Exception e) {
+                logger.error("Exception monitoring FS health", e);
+            }
+        }
+
+        private void monitorFSHealth() {
+            for (Path path : nodeEnv.nodeDataPaths()) {
+                long executionStartTime = currentTimeMillisSupplier.getAsLong();
+                slowPathHealthCheckTimeoutHandler = threadPool.schedule(new Runnable() {
```

Comment:
This spawns (and hopefully cancels) a lot of tasks each iteration. Given that we're not considering timing out here, let's just log at the end of each path's check if it was slow. This has the added advantage that we can report exactly how long it took, not just that it took >5s.

Reply:
The reason why I think the timeout handler works better, is if the node is already stuck on IO we won't get to know till it recovers. Let me know what you think